src/object_cmd.cpp

```diff
@@ -346,21 +348,90 @@ CommandCost CmdBuildObject(DoCommandFlag flags, TileIndex tile, ObjectType type,
 			/* This may never be constructed using this method. */
 			return CMD_ERROR;
 
-		default: // i.e. NewGRF provided.
+		default: { // i.e. NewGRF provided.
+			build_object_size = GB(spec->size, 0, 4) * GB(spec->size, 4, 4);
 			break;
+		}
+	}
+
+	/* Don't allow building more objects if the company has reached its limit. */
+	if (c != nullptr && GB(c->build_object_limit, 16, 16) < build_object_size) {
+		return_cmd_error(STR_ERROR_BUILD_OBJECT_LIMIT_REACHED);
 	}
 
 	if (flags & DC_EXEC) {
 		BuildObject(type, tile, _current_company == OWNER_DEITY ? OWNER_NONE : _current_company, nullptr, view);
 
 		/* Make sure the HQ starts at the right size. */
 		if (type == OBJECT_HQ) UpdateCompanyHQ(tile, hq_score);
+
+		/* Subtract the tile from the build limit. */
+		if (c != nullptr) c->build_object_limit -= 1 << 16;
 	}
 
 	cost.AddCost(ObjectSpec::Get(type)->GetBuildCost() * size_x * size_y);
```

Comment:
I know this is not from this patch, the object spec is available from line 211, so it could be simplified, once we are here :)
```suggestion
	cost.AddCost(spec->GetBuildCost() * size_x * size_y);
```